Update to future intake-esm with polars

aulemahal opened the pull request:
<!-- Please ensure the PR fulfills the following requirements! -->
<!-- If this is your first PR, make sure to add your details to the AUTHORS.rst! -->
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #618
- [x] (If applicable) Documentation has been added / updated (for bug fixes / features).
- [x] (If applicable) Tests have been added.
- [x] This PR does not seem to break the templates.
- [x] CHANGELOG.rst has been updated (with summary of main changes).
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.

### What kind of change does this PR introduce?
The newer intake-esm now uses `polars` to read catalog csv. This drops the support for some of the `read_csv` arguments we were using to enhance performance. While the current intake-esm doesn't fully support what we need, intake/intake-esm#734 fixes that. So the next release should be compatible, with the changes made here.

I also added `intake-esm`'s main in the upstream CI.

### Does this PR introduce a breaking change?
Yes, as the read_csv "converters" are not usable anymore, I dropped the `ensure_new_xrfreq` one that was executed upon the csv read. Which means catalogs using the old xrfreq specs (3H, H, etc) might break further down the line. This could added back by calling it after reading, but it felt unnecessary.

The dtypes of the columns have changed. We can force "categorical" anymore, so everything is a string, except "variable" (an object) and the dates (the way we convert those hasn't changed). I yet have to check if performance is affected. 


### Other information:
As said above, I haven't compared the performance yet.

aulemahal committed:
Update to future intake-esm with polars - remove support for old xrfreq in catalogs

aulemahal committed:
Simplify open with polars dtypes

aulemahal committed:
merge

aulemahal committed:
Add polars to deps

juliettelavoie commented:
Merci Pascal pour les tests! J'ai vu que le gars de intake semble dire qu'on pourra choisir notre backend finalement sur ton issue. J'imagine qu'on peut juste faire ça en attendant que polars soit meilleur en recherche ?

aulemahal committed:
WIP zarr3

aulemahal committed:
merge main

aulemahal committed:
xarray transparent ZarrZip support inside xscen

aulemahal committed:
comment - update get_engine - changelog

aulemahal committed:
Drop Python 3.10 - unsupported by zarr 3

aulemahal committed:
I am so naive

Zeitsperre commented on .github/workflows/main.yml:
```diff
@@ -73,10 +73,6 @@ jobs:
-          - os: 'ubuntu-latest'
-            python-version: "3.10"
-            esmf-version:  "8.7.0"  # 8.8.0 breaks h5py
-            xclim-version: "0.56"  # minimum supported version
```
It might be a good idea to specify at least one build that checks against the minimum support `xclim`.

Zeitsperre commented on tox.toml:
Don't forget to modify this as well

aulemahal committed:
Adapt to zarr 3 - changes

aulemahal committed:
re-add a ci run with min xclim - upd xesmf in ci - fix fastprogress in pyproject

aulemahal committed:
touch ups

aulemahal commented:
@Zeitsperre @juliettelavoie @RondeauG  This PR is now in a mergeable state.

## Zipped Zarrs
I dropped the idea of having a fix for zipped zarr in intake-esm. My PR has been sitting (with a few others) for a while now, I think the core team has had reduced time for maintenance in the last months. Instead of waiting for pushing what feels like a hack anyway, I hacked it here.

In theory, all the modifications below could be removed when the "url pipeline" stuff is implemented in zarr.

`xscen.zarrzip` exposes a new xarray engine that inherits from the zarr one, but wraps zip files in a `ZipStore`, that's all. This means that if xscen is installed, `xr.open_dataset("file.zarr.zip")` will open normally with zarr 3. However, `xr.open_zarr` **can not** be used as this function hardcodes using the default backend. This new "engine" has been added to `xs.io.get_engine`.

Also this means that opening a file with xarray now requires importing xscen because we have all modules imported directly from the top level. If xscen was previously unimported, this means the first call to `xr.open_dataset` takes a few seconds.

Intake-esm has it's own way of inferring the engine, mostly using the "format" column. Without a PR upstream, I had to hack something again. Here, I suggest skipping intake-esm's inference, letting xarray do it. This is done by passing `engine=None` to the xarray open kwargs (intake-esm will only set this if absent). I added this to the default kwargs we pass through `extract_dataset` and `Catalog.to_dataset`.   This means that this will **not** work with `Catalog.to_dataset_dict` and `Catalog.to_dask`.

As a side effect, this allows the use of xarray's `netcdf_engine_order` option by default. Previously, netcdf files where send to `netCDF4` by intake-esm when `engine` was not set. It also removes a lot of the smart code dealing with some remote storage options that intake-esm has. As I wrote in the changelog, I assumed xscen does not aim to support the same range of remote storage options. It is first and foremost aimed to deal with local files, and maybe whatever PAVICS exposes (when w get to the point of having catalogs on there).

## Catalog creation changes
With the switch to polars by intake-esm, I had to change the way we read in some columns. 

Datetime columns can be read by polars, so I changed the default kwargs. I renamed the argument to `read_kwargs`, the same as in intake-esm. In theory this is backend-agnostic, but of course pandas and polars don't have the same arguments names, so the current default expect polars to be used.

Previously, we were parsing the variable column into tuples through some special options to `pd.read_csv`. intake-esm exposed an option for that, so I made use of it (`columns_with_iterables`). Currently, this needs to be given explicitly to intake-esm, which is not great as it depends on the catalog. I suggested to add a field to the ESM collection spec, but this PR also has been sitting unanswered for a while. I implemented the idea here nonetheless. In theory, this could allow for more columns to be iterables.

In any case, I made it now clearer that xscen expects the "variable" column to be filled with tuples. `xs.parse_directory` will make sure it's the case.

Finally, now that reading catalog files is a bit more complicated, I decided to drop the possibility of passing multiples csv files to `Catalog.from_df`. No test were testing this, I am guessing it's not much used.

## Zarr 3
The new intake-esm depends on zarr >=3, so we now do too.  We had a few spots where we were dealing with zarr files manually, which I adapted or moved to use `zarr` instead.

Zarr 3 will read zarr 2 files transparently, but will write files that zarr 2 can't read. Pass `zarr_kwargs={'zarr_format': 2}` to write compatible file in `save_to_zarr`. 

Rechunker does not support zarr 3. I removed the dependency and functions in this PR, but #671 is still open to explore the alternative.  The other way is to take possession of the repo and maintain it.

## Conclusion
I expect this PR to be breaking in places I didn't think of. I think in general  reading (our current catalogs, files) will not be affected, but it is in the writing aspect I worry.

aulemahal commented:
Désolé tous. Ma solution d'un engine xarray custom ne fonctionne pas à tout coup avec la gestion de format/engine que fait intake-esm, du moins pas aussi facilement que je pensais.
